gnmidiff/setrequest.go

```diff
@@ -17,13 +17,183 @@ package gnmidiff
 
 import (
 	"fmt"
+	"sort"
+	"strconv"
+	"strings"
 
 	"github.com/derekparker/trie"
 	"github.com/openconfig/ygot/ygot"
+	"github.com/openconfig/ygot/ytypes"
 
 	gpb "github.com/openconfig/gnmi/proto/gnmi"
 )
 
+// MismatchedUpdate represents two different update values for the same leaf
+// node.
+type MismatchedUpdate struct {
+	// A is the update value in A.
+	A interface{}
+	// B is the update value in B.
+	B interface{}
+}
+
+// SetRequestIntentDiff contains the intent difference between two SetRequests.
+//
+// - The key of the maps is the string representation of a gpb.Path constructed
+// by ygot.PathToString.
+// - The value of the update fields is the JSON_IETF representation of the
+// value.
+type SetRequestIntentDiff struct {
+	AOnlyDeletes      map[string]struct{}
+	BOnlyDeletes      map[string]struct{}
+	CommonDeletes     map[string]struct{}
+	AOnlyUpdates      map[string]interface{}
+	BOnlyUpdates      map[string]interface{}
+	CommonUpdates     map[string]interface{}
+	MismatchedUpdates map[string]MismatchedUpdate
+}
+
+// Format is the string format of any gNMI diff utility.
+type Format struct {
+	// Full indicates that common values are also output.
+	Full bool
+	// TODO: Implement IncludeList and ExcludeList.
+	// IncludeList is a list of paths that will be included in the output.
+	// wildcards are allowed.
+	//
+	// empty implies all paths are included.
+	//IncludeList []string
+	// ExcludeList is a list of paths that will be excluded from the output.
+	// wildcards are allowed.
+	//
+	// empty implies no paths are excluded.
+	//ExcludeList []string
+}
+
+func formatJSONValue(value interface{}) interface{} {
+	if v, ok := value.(string); ok {
+		return strconv.Quote(v)
```

Comment:
It is probably the case, but I need to ask, are we sure that the value was not already quoted?

Reply:
Since the stored JSON values are already unmarshalled into Go types, JSON strings have already been de-quoted before being passed to this function for human-readable formatting.